File: demo/src/main/java/com/umut/petexpandable/SampleDataIndexProvider.java
package com.umut.petexpandable;

import androidx.annotation.NonNull;

import com.umut.expandablrecyclerview.adapter.data.ExpandableDataIndexProvider;
import com.umut.petexpandable.model.MainStreamGenre;
import com.umut.petexpandable.model.SubGenre;

import java.util.List;
import java.util.Map;

public class SampleDataIndexProvider implements ExpandableDataIndexProvider {

    private final List<MainStreamGenre> mainStreamGenres;
    private final Map<MainStreamGenre, List<SubGenre>> subGenres;

    public SampleDataIndexProvider(@NonNull List<MainStreamGenre> mainStreamGenres,
                                   @NonNull Map<MainStreamGenre, List<SubGenre>> subGenres) {
        this.mainStreamGenres = mainStreamGenres;
        this.subGenres = subGenres;
    }

    @Override
    public int getChildrenSize(int parentIndex) {
        return subGenres.get(mainStreamGenres.get(parentIndex)).size();
    }

    @Override
    public int getParentSize() {
        return mainStreamGenres.size();
    }

}

File: demo/src/main/java/com/umut/petexpandable/SampleDataIndexProvider.kt
package com.umut.petexpandable

import com.umut.expandablerecyclerview.item.ListViewItem
import com.umut.expandablerecyclerview.provider.ExpandableDataSource
import com.umut.petexpandable.model.MainStreamGenre
import com.umut.petexpandable.model.SubGenre

class SampleDataIndexProvider(
    private val mainStreamGenres: List<MainStreamGenre>,
    private val subGenres: Map<MainStreamGenre, List<SubGenre>>
) : ExpandableDataSource<SubGenre, MainStreamGenre> {

    override fun getItems(): Iterable<MainStreamGenre> {
        return mainStreamGenres
    }

    override fun <I : ListViewItem> getChildDataSource(listViewItem: ListViewItem): ExpandableDataSource<I, SubGenre>? {
        return object : ExpandableDataSource<I, SubGenre> {
            override fun getItems(): Iterable<SubGenre> {
                return subGenres[listViewItem as MainStreamGenre] ?: emptyList()
            }

            override fun <R : ListViewItem> getChildDataSource(
                listViewItem: ListViewItem
            ): ExpandableDataSource<R, I>? = null
        }
    }
}
